torchao/quantization/prototype/qat/fake_quantizer.py

```diff
@@ -0,0 +1,116 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+
+# This source code is licensed under the license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import Optional, Tuple
+
+import torch
+
+from torchao.quantization.utils import (
+    get_group_qparams_symmetric,
+    get_groupwise_affine_qparams,
+)
+from .api import (
+    FakeQuantizeConfig,
+    QuantizationGranularity,
+)
+from .utils import (
+    _choose_qparams_per_token_asymmetric,
+    _fake_quantize_per_channel_group,
+    _fake_quantize_per_token,
+    _get_qmin_qmax,
+)
+
+
+class FakeQuantizer(torch.nn.Module):
+    """
+    Generic module for applying fake quantization to a tensor, as specified in the config.
+    """
+    def __init__(self, config: FakeQuantizeConfig):
+        super().__init__()
+        self.config = config
+        self.enabled = True
+        self.scale: Optional[torch.Tensor] = None
+        self.zero_point: Optional[torch.Tensor] = None
+
+        # TODO: support range learinng
+        if self.config.range_learning:
+            raise NotImplementedError("Range learning is not supported yet")
+
+    def forward(self, x: torch.Tensor):
+        """
+        Apply fake quantization to the tensor based on the bit-width,
+        granularity, symmetry, and other properties specified in the config.
+        """
+        if not self.enabled:
+            return x
+
+        if self.config.granularity == QuantizationGranularity.PER_TOKEN:
```

Comment:
we have some config here: https://github.com/pytorch/ao/blob/dec03131484b5717e9fb8b6c0c25f41ec30e9fef/torchao/quantization/observer.py#L20 might be good to reuse